File: leetcode_872.py

class TreeNode:
    def __init__(self, left: None, right: None, val=0) -> None:
        self.val = val
        self.left = left
        self.right = right

class Solution:
    leafsRootOne = []
    leafsRootTwo = []

    def preOrderTraverse(self, node: TreeNode, treeNo: int) -> None:
        if node != None:
            if (node.left == None) and (node.right == None):
                # its a leaf
                if treeNo == 0:
                    self.leafsRootOne.append(node.val)
                else:
                    self.leafsRootTwo.append(node.val)

            self.preOrderTraverse(node.left, treeNo)
            self.preOrderTraverse(node.right, treeNo)

    def leafSimilar(self, rootOne: TreeNode, rootTwo: TreeNode) -> bool:
        self.preOrderTraverse(rootOne, 0)
        self.preOrderTraverse(rootTwo, 1)

        # print(self.leafsRootOne)
        # print(self.leafsRootTwo)

        return ''.join(str(x) for x in self.leafsRootOne) == ''.join(str(x) for x in self.leafsRootTwo)


obj = Solution()
rootOne = TreeNode(
        TreeNode(TreeNode(None, None, 4), TreeNode(None, None, 5), 2),
        TreeNode(TreeNode(None, None, 6), TreeNode(None, None, 7), 3),
        1
    )

rootTwo = TreeNode(
        TreeNode(TreeNode(None, None, 4), TreeNode(None, None, 5), 2),
        TreeNode(TreeNode(None, None, 6), TreeNode(None, None, 8), 3),
        1
    )

res = obj.leafSimilar(rootOne, rootTwo)
print(res)
        

File: leetcode_872.js
// Leet code 872 Similar Leaf
// Oct 09 2023 By Nahid Hasan Khan

/**
 * Definition for a binary tree node.
 */

class TreeNode {
    constructor(val = 0, left = null, right = null) {
        // this.val = (val === undefined ? 0 : val);
        // this.left = (left === undefined ? null : left);
        // this.right = (right === undefined ? null : right);
        this.val = val;
        this.left = left;
        this.right = right;
    }

    static printTree = (node) => {
        if (node !== null) {
            console.log(node.val);
            this.printTree(node.left);
            this.printTree(node.right);
        }
    }
}

/** 
 * Algorithm recursive pre order tree traverse
 * Check if the node has no left and right node, means a leaf
 * Collect all the leafs for tree 1 and tree 2 and later match the string created
 *  from those leafs
*/

/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function(root1, root2) {
    let leafsOne = [];
    let leafsTwo = [];

    // traverse the trees and collect the leaf nodes
    preOrderTraverse(root1, leafsOne);
    preOrderTraverse(root2, leafsTwo);

    console.log(leafsOne);
    console.log(leafsTwo);

    // check if the two array have same contents
    if (leafsOne.length != leafsTwo.length) return false;

    for (var i = 0; i < leafsOne.length; i++) {
        if (leafsOne[i] != leafsTwo[i]) return false;
    }

    return true;
};

const preOrderTraverse = (node, leafs) => {
    if (node !== null) {
        if (node.left == null && node.right == null) {
            leafs.push(node.val);
        }

        preOrderTraverse(node.left, leafs);
        preOrderTraverse(node.right, leafs);
    }
}

const root1 = new TreeNode(1, new TreeNode(2, null, null), new TreeNode(4));
const root2 = new TreeNode(1, new TreeNode(2, null, null), new TreeNode(3, null, null));

console.log(leafSimilar(root1, root2));
// TreeNode.printTree(root1);